### DISCUSSION TITLE
Repository history and size

### DISCUSSION THREAD

**Discussion by landsman** [storybookjs/storybook#24289] [Documentation feedback]
Guys, I wanted to contribute on the doc, which I wasn't able to because of [this issue](https://github.com/storybookjs/storybook/issues/24288) anyway there is another topic that I want to mention.

## Repository size, history

**It's huge!** I did a fresh git clone of the repository and there are some numbers:

1. The size of the raw repository is **831,3 MB** right now
2. From this is **768,8 MB** just .git folder
3. There are [53,265 commits](https://github.com/storybookjs/storybook/commits/next) in the main branch
5. There are **1700+ issues** for several packages
6. There are 127 pull requests

## Questions

Is this maintainable in the future? What about splitting it into more repositories for smaller packages and parts?

- easier to read and maintain issues
- smaller git repository size
- easier to contribute

Let's be honest - who is going to search among all these issues for something related to integration and answer it? Who will voluntarily clone such a huge repository and read such a huge git log every time?

I really like this project and the current state makes me think about these questions.

**Comment by shilman** [comment]
@landsman storybook is a big project and will always have a lot of issues because of its surface area/broad compatibility. I think we can get it down to 1000 but probably not much less.

Regarding repo size, perhaps we can archive the repo and start over from v7 or something to preserve some limited history. I think a lot of the large size is due to older versions where we included React Native, which has been factored out into its own repo years ago. @storybookjs/core WDYT?

**Reply by JReinhold** [reply]
Sounds good. My suggestion would be to split the history in two, anything before v6.5.16 goes to an "archive repo".
When you want an effective git blame, you can apparently temporarily merge the repos together again, locally.


https://andrewlock.net/reducing-the-size-of-a-git-repository-with-git-replace/

**Reply by IanVS** [reply]
Another thing that could be done is to use squash merges for PRs.  Often the git history ends up pretty messy because we have a lot of commits like "fix tests" or similar.  I think granular commits can be useful in some cases, but I'm not sure we have the commit discipline to make them truly useful at that granularity.  And if someone needs to dig deeper than the squashed PR commit, there's always the GitHub PR to look through.  This could reduce the speed at which the `.git` history grows, since instead of sometimes 50 commits landing in a PR, it would just be one.

**Reply by yannbf** [reply]
Storybook is big in size as it contains years of work, all of which are important and can be useful at one point. It seems like we should just suggest new contributors to shallow clone the repo instead? cc @kylegach

**Reply by ndelangen** [reply]
Another mistake we made in the past was having a bunch of larger images and even videos and gifs in the repository.

We could squash a bunch of history that's really old.
Or start a new repository...
Both of those options are disruptive. It can cause existing branches to become unmergable due to git refusing to merge unrelated histories. (which can be overcome)

I'm not convinced this is a important enough problem to be tackling right now.
Users/contributors could maybe do a shallow checkout if they wish..

**Reply by landsman** [reply]
@IanVS yes! It can be done by default in GitHub repository settings. I did this on all my repositories to make them clean. History can be found in the linked PR link anyway.

@yannbf not everything is relevant for them I guess. Two years old commits are relevant for very few people in projects like this. 

@ndelangen oh, that's it probably! It can be removed I even in history I think. I did these historical removal in few repositories by tool like this one: https://rtyley.github.io/bfg-repo-cleaner/

**Reply by IanVS** [reply]
@yannbf FWIW, GitHub recommends against shallow clones:

>These clones truncate the commit history to reduce the clone size. This creates some unexpected behavior issues, limiting which Git commands are possible. These clones also put undue stress on later fetches, so they are strongly discouraged for developer use. They are helpful for some build environments where the repository will be deleted after a single build.

https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone

I wonder if recommending that contributors use that could cause other headaches.

From that article, it seems that a _blobless_ clone is what we might want.  `git clone --filter=blob:none <url>`.  I'm going to try running that way for a bit.  The clone itself only pulled in `63.96 MiB` and was much faster than the full clone, so off to a good start so far.

**Reply by landsman** [reply]
making things as least as uncomplicated as possible should be the goal if there is interest in occasional contributors so that they are not afraid to get involved in my opinion

**Reply by ndelangen** [reply]
@landsman I bet rewriting history would be very disruptive to all existing branches & PRs, wouldn't it @landsman ?

**Reply by landsman** [reply]
> @landsman I bet rewriting history would be very disruptive to all existing branches & PRs, wouldn't it @landsman ?

not necessarily, try doing it locally and check for conflicts / changed git commits sha

**Reply by JReinhold** [reply]
@ndelangen If we do the cutoff point at v6.5.X then surely a minimal amount of open branches/PRs would be affected? I'd argue that they should probably be redone anyway.

**Reply by ndelangen** [reply]
@JReinhold I would not be directly opposed, is there an easy way for us to find out, how to do this, without actually doing it on the main repo?
